contracts/metatx/MinimalForwarder.sol

```
@@ -26,29 +21,64 @@ contract MinimalForwarder is EIP712 {
+    constructor() EIP712(name(), version()) {}
+
+    function version() public view virtual returns (string memory) {
+        return "0.0.1";
+    }
 
-    function getNonce(address from) public view returns (uint256) {
+    function name() public view virtual returns (string memory) {
+        return "MinimalForwarder";
+    }
+
+    function getNonce(address from) public view virtual returns (uint256) {
         return _nonces[from];
     }
 
-    function verify(ForwardRequest calldata req, bytes calldata signature) public view returns (bool) {
-        address signer = _hashTypedDataV4(
-            keccak256(abi.encode(_TYPEHASH, req.from, req.to, req.value, req.gas, req.nonce, keccak256(req.data)))
-        ).recover(signature);
-        return _nonces[req.from] == req.nonce && signer == req.from;
+    function verify(ForwardRequest calldata req, bytes calldata signature) public view virtual returns (bool) {
+        return _verify(_toExpiring(req), _recoverSigner(req, signature));
+    }
+
+    function verify(ExpiringForwardRequest calldata req, bytes calldata signature) public view virtual returns (bool) {
+        return _verify(req, _recoverSigner(req, signature));
     }
 
     function execute(
         ForwardRequest calldata req,
         bytes calldata signature
-    ) public payable returns (bool, bytes memory) {
+    ) public payable virtual returns (bool, bytes memory) {
         require(verify(req, signature), "MinimalForwarder: signature does not match request");
+        return _execute(_toExpiring(req));
+    }
+
+    function execute(
+        ExpiringForwardRequest calldata req,
+        bytes calldata signature
+    ) public payable virtual returns (bool, bytes memory) {
+        require(verify(req, signature), "MinimalForwarder: signature does not match request");
+        return _execute(req);
+    }
+
+    function _execute(ExpiringForwardRequest memory req) private returns (bool, bytes memory) {
```

Is the fact that `_execute` is not reverting if the call fails intended and left to the contract inheriting from `MinimalForwarder`?

If yes and given that this version will be considered as production ready, it would be nice to add a warning about this to avoid silent failure on the implementation side. Else, I think we should revert while bubbling up the error.

---

Yes, its an intended behavior.

This PR is on hold. The code it contains should not be seen as a production-ready version of the forwarder. For sure we will want to challenge that design choice (IMO there are good reason for this behavior), and document it.

---

Clear, thx for the answer.

---

➕  to @Amxx explanation.

I'll leave this open and I added a note at the top of the PR indicating this is currently on hold.